### PR TITLE
feat(connector): implement authorize and capture flows for Fiserv

### LINKED ISSUE #229: Add Fiserv for card payments

### PR DESCRIPTION
**Pull request by zereraz**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
Authorize and capture flow
Todo: Issue for optimising the encoding process in header and body #370 

### Additional Changes

- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Adding a new connector `Fiserv`

Closes #229.

## How did you test it?
Added unit tests for mentioned payment flows.

Screenshot 2023-01-13 at 7 43 38 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/fiserv.rs [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+            .duration_since(UNIX_EPOCH)
+            .into_report()
+            .change_context(errors::ConnectorError::RequestEncodingFailed)?
+            .as_millis()
+            .to_string();
+        let auth: fiserv::FiservAuthType =
+            fiserv::FiservAuthType::try_from(&req.connector_auth_type)?;
+        let api_key = auth.api_key.clone();
+
+        let fiserv_req = utils::Encode::<fiserv::FiservCaptureRequest>::convert_and_encode(req)
```
Comment (jarnura) [inline crates/router/src/connector/fiserv.rs:125]: we can call the `Self::get_request_body` here

File: crates/router/src/connector/fiserv.rs [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+            .duration_since(UNIX_EPOCH)
+            .into_report()
+            .change_context(errors::ConnectorError::RequestEncodingFailed)?
+            .as_millis()
+            .to_string();
+        let auth: fiserv::FiservAuthType =
+            fiserv::FiservAuthType::try_from(&req.connector_auth_type)?;
+        let api_key = auth.api_key.clone();
+
+        let fiserv_req = utils::Encode::<fiserv::FiservPaymentsRequest>::convert_and_encode(req)
```
Comment (jarnura) [inline crates/router/src/connector/fiserv.rs:280]: we can use `Self::get_request_body` here

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/fiserv/test.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
+use futures::future::OptionFuture;
```
[the remark this answers is not in the document]
Comment (zereraz) [reply crates/router/src/connector/fiserv/test.rs:1]: I think these got autogenerated. No I didn't run them.

File: crates/router/src/connector/fiserv.rs [hunk 4]
```diff
@@ -112,26 +110,17 @@ impl
         req: &types::PaymentsCaptureRouterData,
         _connectors: &settings::Connectors,
     ) -> CustomResult<Vec<(String, String)>, errors::ConnectorError> {
-        let timestamp = SystemTime::now()
-            .duration_since(UNIX_EPOCH)
-            .into_report()
-            .change_context(errors::ConnectorError::RequestEncodingFailed)?
-            .as_millis()
-            .to_string();
+        let timestamp = time::OffsetDateTime::now_utc().unix_timestamp();
```
Comment (SanchithHegde) [inline crates/router/src/connector/fiserv.rs:113]: Sorry I forgot to mention this earlier, could you please add a utility function in the `common_utils::date_time` module called `now_unix_timestamp()` which returns this value?

**Review by SanchithHegde — DISMISSED** [review]

File: crates/router/src/connector/fiserv/transformers.rs [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+                let merchant_details = MerchantDetails {
+                    merchant_id: auth.merchant_account,
+                    terminal_id: "10000001".to_string(),
+                };
+
+                let transaction_interaction = TransactionInteraction {
+                    origin: "ECOM".to_string(),
+                    eci_indicator: "CHANNEL_ENCRYPTED".to_string(),
+                    pos_condition_code: "CARD_NOT_PRESENT_ECOM".to_string(),
+                };
```
[the remark this answers is not in the document]
Comment (SamraatBansal) [reply crates/router/src/connector/fiserv/transformers.rs]: Added Comments to these fields to provide context

**Comment by SanchithHegde** [comment]
@SamraatBansal Please run the tests and attach a screenshot in the PR description.

**Review by jarnura — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**